src/azstoragetorch/datasets.py

```diff
@@ -106,3 +92,93 @@ def __getitem__(self, index: int) -> _TransformOutputType_co:
 
     def __len__(self) -> int:
         return len(self._blobs)
+
+
+class IterableBlobDataset(torch.utils.data.IterableDataset[_TransformOutputType_co]):
+    def __init__(
+        self, blobs: Iterable[Blob], transform: _TRANSFORM_TYPE = _default_transform
+    ):
+        self._blobs = blobs
+        self._transform = transform
+
+    @classmethod
+    def from_blob_urls(
+        cls,
+        blob_urls: Union[str, Iterable[str]],
+        *,
+        credential: _client.AZSTORAGETORCH_CREDENTIAL_TYPE = None,
+        transform: _TRANSFORM_TYPE = _default_transform,
+    ) -> "IterableBlobDataset":
```

Comment:
Sorry I didn't catch this before in other places but is `typing.Self` a better return type here? I think `typing.Self` does a better job with inheritance typing. [PEP](https://peps.python.org/pep-0673/)

It is Python 3.11 though so we either would need `typing_extensions` and/or a conditional import.

Reply:
That's a good callout. I do think `typing.Self` is better a return type as well. I also think it handles generics too, which I think got missed in this return type (e.g. `IterableBlobDataset[_TransformOutputType_co]`)

In terms of adding `typing_extensions`, that sounds good to me. The dependency is already in the library's tree since the SDKs use it. I've also seen other instances where it would have been nice to be able to use some typing backports.

I'm going to merge this PR as is and then send a follow up PR that adds `typing_extensions` to the project and do a pass of updating the types for datasets.